ignite/pkg/cosmostxcollector/adapter/postgres/postgres.go

```diff
@@ -0,0 +1,351 @@
+	schemas                        Schemas
+}
+
+// UpdateSchema updates the database schema to the latest version available.
+// It applies all available schemas that were not applied already.
+func (a Adapter) UpdateSchema(ctx context.Context, s Schemas) error {
+	db, err := a.getDB()
+	if err != nil {
+		return err
+	}
+
+	// Create the schema table if it doesn't exists
+	if _, err := db.ExecContext(ctx, s.GetTableDDL()); err != nil {
+		return fmt.Errorf("failed to check schema table: %w", err)
+	}
+
+	// Get the current schema version
+	var v uint64
+	if err := db.QueryRowContext(ctx, s.GetSchemaVersionSQL()).Scan(&v); err != nil {
+		return fmt.Errorf("failed to read current schema version: %w", err)
+	}
+
+	return s.WalkFrom(v+1, func(version uint64, script []byte) error {
+		if _, err := db.ExecContext(ctx, string(script)); err != nil {
+			return fmt.Errorf("error applying schema version %d: %w", version, err)
+		}
+
+		return nil
+	})
+}
+
+func (a Adapter) GetType() string {
+	return adapterType
+}
+
+func (a Adapter) Init(ctx context.Context) error {
+	return a.UpdateSchema(ctx, a.schemas)
+}
+
+func (a Adapter) Save(ctx context.Context, txs []cosmosclient.TX) error {
+	db, err := a.getDB()
+	if err != nil {
+		return err
+	}
+
+	// Start a transaction
+	sqlTx, err := db.BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+
+	// Rollback won't have any effect if the transaction is committed before
+	defer sqlTx.Rollback()
+
+	// Prepare insert statements to speed up "bulk" saving times
+	txStmt, err := sqlTx.PrepareContext(ctx, sqlInsertTX)
+	if err != nil {
+		return err
+	}
+
+	defer txStmt.Close()
+
+	evtStmt, err := sqlTx.PrepareContext(ctx, sqlInsertEvent)
+	if err != nil {
+		return err
+	}
+
+	defer evtStmt.Close()
+
+	attrStmt, err := sqlTx.PrepareContext(ctx, sqlInsertEventAttr)
+	if err != nil {
+		return err
+	}
+
+	defer attrStmt.Close()
+
+	// All the transactions are saved within the context of the same database
+	// transactions and because of that either all block transactions are
+	// saved or none of them.
+	for _, tx := range txs {
+		if err := saveRawTX(ctx, sqlTx, tx.Raw); err != nil {
+			return err
+		}
+
+		if err := saveTX(ctx, txStmt, evtStmt, attrStmt, tx); err != nil {
+			return err
+		}
+	}
+
+	return sqlTx.Commit()
+}
+
+func (a Adapter) GetLatestHeight(ctx context.Context) (height int64, err error) {
+	db, err := a.getDB()
+	if err != nil {
+		return 0, err
+	}
+
+	row := db.QueryRowContext(ctx, sqlSelectBlockHeight)
+	if err = row.Scan(&height); err != nil {
+		return 0, err
+	}
+
+	return height, nil
+}
+
+func (a Adapter) Query(ctx context.Context, q query.Query) (query.Cursor, error) {
+	db, err := a.getDB()
+	if err != nil {
+		return nil, err
+	}
+
+	query, err := parseQuery(q)
+	if err != nil {
+		return nil, err
+	}
+
+	args := extractQueryArgs(q)
+	rows, err := db.QueryContext(ctx, query, args...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &cursor{rows}, nil
```

Comment:
@jeronimoalbi if you return `rows` directly, that compiles. So I wonder if this `cursor` type is really useful. WDYT?

Reply:
You are right, I will change it